### ISSUE TITLE
Segmentation fault (11) after avgDistinct(Decimal)

### ISSUE THREAD

**Issue by NikitaZakharov** [ClickHouse/ClickHouse#22020]
ClickHouse receives Segmentation fault (11) error and crashes after simple query using avgDistinct function for decimal type.

ClickHouse version: 20.8.7.15

```sql
SELECT 
  avgDistinct(toDecimal64(number, 4))
FROM numbers(10)
```

**Stacktrace**

```
ch4-amd :) SELECT avgDistinct(toDecimal64(number, 4)) FROM numbers(10);

SELECT avgDistinct(toDecimal64(number, 4))
FROM numbers(10)

[ch4-amd] 2021.03.23 10:25:21.979584 [ 94425 ] <Fatal> BaseDaemon: ########################################
[ch4-amd] 2021.03.23 10:25:21.979698 [ 94425 ] <Fatal> BaseDaemon: (version 20.8.7.15 (official build), build id: 0E6D50A0B39845B6) (from thread 11237) (query_id: 73a82afa-04e6-4854-940c-d66caa7800a3) Received signal Segmentation fault (11)
[ch4-amd] 2021.03.23 10:25:21.979734 [ 94425 ] <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
[ch4-amd] 2021.03.23 10:25:21.979778 [ 94425 ] <Fatal> BaseDaemon: Stack trace: 0xe7d9590 0x15955a70 0x1596a030 0x16544629 0x1654a219 0x16309f00 0x1630ee44 0x1630f73d 0x163101f1 0x163172bd 0xe641b47 0xe640093 0x7f9d61c6c609 0x7f9d61b82293
[ch4-amd] 2021.03.23 10:25:21.979859 [ 94425 ] <Fatal> BaseDaemon: 3. DB::IAggregateFunctionDataHelper<DB::AggregateFunctionAvgData<DB::Decimal<__int128>, unsigned long>, DB::AggregateFunctionAvg<DB::Decimal<long>, DB::AggregateFunctionAvgData<DB::Decimal<__int128>, unsigned long> > >::create(char*) const @ 0xe7d9590 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.979930 [ 94425 ] <Fatal> BaseDaemon: 4. DB::Aggregator::createAggregateStates(char*&) const @ 0x15955a70 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.979963 [ 94425 ] <Fatal> BaseDaemon: 5. DB::Aggregator::executeOnBlock(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >, unsigned long, DB::AggregatedDataVariants&, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> >&, std::__1::vector<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> >, std::__1::allocator<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> > > >&, bool&) @ 0x1596a030 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980001 [ 94425 ] <Fatal> BaseDaemon: 6. DB::AggregatingTransform::consume(DB::Chunk) @ 0x16544629 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980030 [ 94425 ] <Fatal> BaseDaemon: 7. DB::AggregatingTransform::work() @ 0x1654a219 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980059 [ 94425 ] <Fatal> BaseDaemon: 8. ? @ 0x16309f00 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980086 [ 94425 ] <Fatal> BaseDaemon: 9. ? @ 0x1630ee44 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980103 [ 94425 ] <Fatal> BaseDaemon: 10. DB::PipelineExecutor::executeImpl(unsigned long) @ 0x1630f73d in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980130 [ 94425 ] <Fatal> BaseDaemon: 11. DB::PipelineExecutor::execute(unsigned long) @ 0x163101f1 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980156 [ 94425 ] <Fatal> BaseDaemon: 12. ? @ 0x163172bd in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980187 [ 94425 ] <Fatal> BaseDaemon: 13. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xe641b47 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980213 [ 94425 ] <Fatal> BaseDaemon: 14. ? @ 0xe640093 in /usr/bin/clickhouse
[ch4-amd] 2021.03.23 10:25:21.980236 [ 94425 ] <Fatal> BaseDaemon: 15. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
[ch4-amd] 2021.03.23 10:25:21.980278 [ 94425 ] <Fatal> BaseDaemon: 16. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so

```

**Comment by kitaisreal** [comment]
@NikitaZakharov could not reproduce on current master.
```
SELECT 
  avgDistinct(toDecimal64(number, 4))
FROM numbers(10)
```
```
┌─avgDistinct(toDecimal64(number, 4))─┐
│                                 4.5 │
└─────────────────────────────────────
```

**Comment by alexey-milovidov** [comment]
> 20.8.7

Version 20.8.7 is not the latest patch release.

**Comment by lirong9392** [comment]
> > 20.8.7
> 
> Version 20.8.7 is not the latest patch release.

**Comment by lirong9392** [comment]
> > 20.8.7
> 
> Version 20.8.7 is not the latest patch release.

Hello，I also encountered a similar stack on the 21.3 version. Is this problem fixed on the 21.3 version?

````
2023.05.29 09:03:45.939672 [ 35252 ] {} <Fatal> BaseDaemon: ########################################
2023.05.29 09:03:45.939854 [ 35252 ] {} <Fatal> BaseDaemon: (version 21.3.4.25, no build id) (from thread 109070) (query_id: 804cd7c4-c93b-4932-8c80-e34733c63757) Received signal Segmentation fault (11)
2023.05.29 09:03:45.939888 [ 35252 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2023.05.29 09:03:45.939932 [ 35252 ] {} <Fatal> BaseDaemon: Stack trace: 0x55c19412cbe4 0x55c1996279d0 0x55c19967764f 0x55c199ab3773 0x55c199ab9a14 0x55c1992da27d 0x55c1992e0d97 0x55c1992e1926 0x55c192f620f3 0x55c192f619bf 0x7fccbef56dd5 0x7fccbec7fead
2023.05.29 09:03:45.961707 [ 35252 ] {} <Fatal> BaseDaemon: 3. DB::IAggregateFunctionDataHelper<DB::AggregateFunctionSumData<DB::Decimal<__int128> >, DB::AggregateFunctionSum<DB::Decimal<long>, DB::Decimal<__int128>, DB::AggregateFunctionSumData<DB::Decimal<__int128> >, (DB::AggregateFunctionSumType)0> >::create(char*) const @ 0x5a83be4 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.962314 [ 35252 ] {} <Fatal> BaseDaemon: 4. DB::Aggregator::createAggregateStates(char*&) const @ 0xaf7e9d0 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.962838 [ 35252 ] {} <Fatal> BaseDaemon: 5. DB::Aggregator::executeOnBlock(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >, unsigned long, DB::AggregatedDataVariants&, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> >&, std::__1::vector<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> >, std::__1::allocator<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> > > >&, bool&) @ 0xafce64f in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.963357 [ 35252 ] {} <Fatal> BaseDaemon: 6. DB::AggregatingTransform::consume(DB::Chunk) @ 0xb40a773 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.963890 [ 35252 ] {} <Fatal> BaseDaemon: 7. DB::AggregatingTransform::work() @ 0xb410a14 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.964406 [ 35252 ] {} <Fatal> BaseDaemon: 8. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0xac3127d in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.964928 [ 35252 ] {} <Fatal> BaseDaemon: 9. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) (.constprop.1) @ 0xac37d97 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.965426 [ 35252 ] {} <Fatal> BaseDaemon: 10. ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long)::'lambda0'()>(DB::PipelineExecutor::executeImpl(unsigned long)::'lambda0'()&&)::'lambda'()::operator()() @ 0xac38926 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.965946 [ 35252 ] {} <Fatal> BaseDaemon: 11. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x48b90f3 in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.966476 [ 35252 ] {} <Fatal> BaseDaemon: 12. void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0x48b89bf in /opt/clickhouse/bin/clickhouse
2023.05.29 09:03:45.966518 [ 35252 ] {} <Fatal> BaseDaemon: 13. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
2023.05.29 09:03:45.966560 [ 35252 ] {} <Fatal> BaseDaemon: 14. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
2023.05.29 09:03:46.743712 [ 35252 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 56454E8FC8DC5BE3E7B3B507D44B58C5. There is no information about the reference checksum.
````

my table is:
````
CREATE TABLE OLAP_SUM_CK_CLUSTER( 
	CLUSTER_NUM String COMMENT '集群编号',
	CLUSTER_NAME String COMMENT '集群名',
	NAMESPACE_COUNT Int64 COMMENT '应用个数',
	TABLE_CNT Int64 COMMENT '总表个数',
	FLINK_CNT Int64 COMMENT '实时表个数', 
	SELECT_COUNT_ALL Int64 COMMENT '资询SQL', 
	INSERT_COUNT_All Int64 COMMENT '插入SQL',
	SELECT_ERROR_COUNT Int64 COMMENT '查询报错SQL',
	INSERT_ERROR_COUNT Int64 COMMENT '插入报错SQL',
	JOB_COUNT Int64 COMMENT '作业个数',
	JOB_ERROR_COUNT Int64 COMMENT '报错作业个数', 
	DISK_All Decimal(11,2) COMMENT '总存储空间', 
	DISK_USE Decimal(11,2) COMMENT '已用存储空间', 
	MEMOPY_ALL Decimal(11,2) COMMENT '总内存',
	MEMORY_MAX_RATE Decimal(11, 2) COMMENT '最大内存使用车', 
	MEMORY_AVG_RATE Decimal(11, 2) COMMENT '平均内存使用塞', 
	CPU_MAX_RATE Decimal(11, 2) COMMENT '最大ceu使用率',
	PARALLEL_MAX Decimal(11,2) COMMENT '最大并发',
	CPU_AVG_RATE Decimal (11,2) COMMENT '平均cpu使用态',
	PARALLEL_AVG Decimal(11,2) COMMENT '平均并发',
	DW_SNSH_DT Date COMMENT '快照日期'
)
ENGINE = ReplicatedMergeTree('/clickhouse/default_cluster/OLAP/OLAP_SUM_CK_CLUSTBR/{shard}', '(replica)')
PARTITION BY toYYYYMMDD(DW_SNSH_DT)
ORDER BY (CLUSTER_NUM, CLUSTER_NAME);
````

my sql is：
````
select 
(SUMDistinct(multiIf(OLAP_SUM_CK_CLUSTER.DW_SNSH_DT < '2023-05-29', OLAP_SUM_CK_CLUSTER.MEMORY_AVG_RATE, 0)) - SUMDistinct(multiIf(OLAP_SUM_CK_CLUSTER.DW_SNSH_DT < addDays(CAST(***,***), -7), 
OLAP_SUM_CK_CLUSTER.MEMORY_AVG_RATE, 0))) / SUMDistinct(multiIf(OLAP_SUM_CK_CLUSTER.DW_SNSH_DT < addDays(CAST(***,***), -7),
OLAP_SUM_CK_CLUSTER.MEMORY_AVG_RATE, 0)) AS d7dfa48a8277484d
FROM OLAP_LOCAL.OLAP_SUM_CK_CLUSTER AS OLAP_SUM_CK_CLUSTER LIMIT 100;
````

**Comment by alexey-milovidov** [comment]
@lirong9392, upgrade to the latest version **immediately**.
